keps/sig-apps/2255-pod-cost/README.md

```diff
@@ -55,102 +48,130 @@ Items marked with (R) are required *prior to targeting to a milestone / release*
 
 ## Summary
 
-This feature allows making a suggestion to the ReplicaSet controller, which pod of a Deployment should be deleted first when a scale-down event happens. This can prevent session disruption in stateful applications in a trivial manner.
+This feature allows applications to give a hint to the ReplicaSet controller 
+as to which pods should be deleted first on scale down.
 
 ## Motivation
 
-For some applications, it is necessary that the application can tell Kubernetes which pod can be deleted and which replica has to be protected. The reason for this is that some applications do have stateful sessions and it is not possible to put such an application into Kubernetes because of session termination resulting from "random" down-scale. If the application is able to tell Kubernetes which of the replicas contains no/few/less important active sessions, this would solve many problems. This feature is non-disruptive to the default behaviour. Only if the annotation is existing, it will make a difference in deletion order.
+Currently ReplicaSets are scaled down based on a criteria that on the
+limit prioritizes deleting pods with a more recent creation/readiness 
+timestamp. This is not ideal for some applications where the cost of 
+deleting pods is not related to how recent they were created. 
 
 ### Goals
 
-To recommend which pod gets deleted next of a ReplicaSet. This should help to avoid major reworks in existing applications architecture:
-* [45509](https://github.com/kubernetes/kubernetes/issues/45509) - Scale down a deployment by removing specific pods
+- Allow applications to influence the order of deleting pods when scaling down a ReplicaSet
 
 
 ### Non-Goals
 
-Guaranteed (in contrast to the recommendation stated in Goals) deletion of a selected replica.
+- Guarantees on pod deletion order
 
 ## Proposal
 
-The application can set the `controller.kubernetes.io/pod-cost` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
+Define a known annotation, namely `controller.kubernetes.io/pod-cost` that 
+applications can set to offer a hint on the cost of deleting a pod compared
+to other pods belonging to the same ReplicaSet. 
 
-If all pods have the same priority, there is no difference in the normal pod delete decision behaviour. The same applies if the pod-cost annotation is not used at all.
+The pod-cost range will be from [-MaxInt, MaxInt]. The default value is 0.
+Invalid values (like setting the annotation to string) will be converted to
+the default value.
 
-The pod-cost annotation can be changed during operation, for example, if workload changes or a new master gets elected.
+When scaling down a ReplicaSet, controller-manager will prioritize deleting
```

Comment:
Shouldn't this be the opposite?

That is, Pods without the annotation get pod-cost 0, thus lowest priority to be removed first. My mental interpretation is that a Pod that has a high cost should be deleted first.

Unless you are going with the interpretation "cost of losing the Pod". But then this confusion suggests that we need a better name for the annotation.

Reply:
What about `pod-delete-cost` or `pod-downscale-cost`?

Reply:
there is another comment discussing the same issue, lets discuss there in one place.

Reply:
actually lets have the discussion here since this where we are defining the semantics.

Reply:
Yes, my interpretation is the cost of deleting the pod. 

The default cost of 0 means by default the cost is in the middle of the range (like we did with priority).

I guess there is confusion about the name, so indeed it is better if we change it. How about `delete-cost`.

@ingvagabund @drbugfinder-work lets have the discussion here where the semantics of the annotation is defined.

Reply:
About the 0 default: if the controller that manages the annotation didn't have time to look at the new Pod, this is somewhat similar to the previous heuristic of "pod is just starting up", so safer to delete.

The opposite approach would be: the controller didn't have time to qualify this new Pod, so we better not remove it yet.

I prefer the first interpretation.

Reply:
The cost is relative to other pods, and so having the default value in the middle allows controllers to select what interpretation they want. They can use positive values for pod cost if they want the first interpretation (which is the typical scenario), or they can use negative values to indicate that pods processed by the controller should always be deleted before the ones that didn't (second interpretation).

`pod-delete-cost` sounds good to me, which aligns with the KEP title.

Reply:
I think that we should not define any default value, as this should be a relative ranking which pod should be deleted first. The word "cost" is also a little bit confusing, as we all have noticed. I would rather describe it as "pod-delete-ranking" and keep the order 
smaller-value=higher-keep-priority=lower-delete-chance

greater-value=lower-keep-priority=higher-delete-chance 

I would say that the range from 0 to +MAXINT should be used for this (0 to -MININT is accepted as well).

Reply:
> I think that we should not define any default value, as this should be a relative ranking which pod should be deleted first. 

We need to define the behavior for the case where some pods don't have the value set and some do in the same ReplicaSet. So defining a default value is necessary, otherwise we can't compare the pods that have the field set and ones that don't.

> The word "cost" is also a little bit confusing, as we all have noticed. I would rather describe it as "pod-delete-ranking" and keep the order

don't you think `pod-delete-cost` clarifies it? rank is similar to priority, a higher rank may give the impression that the pod is more important, and hence should be deleted last, so one can argue that it is just as confusing.

> smaller-value=higher-keep-priority=lower-delete-chance

I think this is tricker to use by applications. For example, in the utilization story, pods with *higher* utilization or more open sessions are preferred to be deleted last. Do you have an example where doing the opposite (smaller->lower chance) fits better?

Reply:
> We need to define the behavior for the case where some pods don't have the value set and some do in the same ReplicaSet. So defining a default value is necessary, otherwise we can't compare the pods that have the field set and ones that don't.
We do not need a default. If the priority is low-value < high-value < not-annotated, always the not-annotated pod(s) get deleted. Of course one can remove an annotation from a prioritized pod and it will be in the group of pods to be deleted first (without any order/ranking).
Please see my demo https://www.youtube.com/watch?v=91APTsEsxIU

>  rank is similar to priority, a higher rank may give the impression that the pod is more important, and hence should be deleted last, so one can argue that it is just as confusing.
I was thinking of a "winners podium" when I read "rank", so for me 1st position would be the highest rank.

> I think this is tricker to use by applications. For example, in the utilization story, pods with higher utilization or more open sessions are preferred to be deleted last. Do you have an example where doing the opposite (smaller->lower chance) fits better?
I am with you at this point - if you see it as a "session counter" then it is a valid order. I'm fine with both options, we just need to define it, so there is no confusion about this anymore.

Reply:
I am inclined to `pod-delete-cost`. It clearly expresses the intention without specifying how the goal is achieved. A word `rank` does not say how the value is going to be perceived. Depending on context it can mean both higher/lower rank -> sooner to delete.

Although, `pod-deletion-cost` sounds a bit better to me. Just a nit. Not sure if there's a generic rule for how verbs in annotation keys are to be bend.

Reply:
Wrt. default value, we need one to assume if there's no annotation set. `0` will make a trouble if there are annotations with negative values. But again, it's up to a component annotating pods to take it into account. The component might just (re-)annotate any pod older than some duration with negative number so pods that are "young" enough and don't have an annotation to be given a chance to still run and thrive.

Reply:
That's why I agree with Abdullah that having the default in the center (0) is the best approach. Users can assign the interpretation they want by adding positive or negative cost.

Reply:
ok, unless I miss-understood the comments, I see that there is agreement to use `pod-delete-cost` and the semantics is the cost of deleting the pod (i.e., lower value means sooner to delete)

Regarding the default value, drbugfinder-work@ the way you are describing it is basically giving -MaxInt default value.

> 0 will make a trouble if there are annotations with negative values.

I think a default of 0 offers flexibility as I argued in https://github.com/kubernetes/enhancements/pull/2272#discussion_r558557145; this is similar to pod priority. It is too restrictive to fix it at either end of the range.